cpp/src/parquet/page_index.cc

```diff
@@ -0,0 +1,174 @@
+  }
+
+  const std::vector<T>& min_values() const override { return min_values_; }
+
+  const std::vector<T>& max_values() const override { return max_values_; }
+
+  const std::vector<int32_t> GetNonNullPageIndices() const override {
+    return non_null_page_indices_;
+  }
+
+ private:
+  /// Wrapped thrift column index.
+  const format::ColumnIndex column_index_;
+  /// Decoded typed min/max values. Null pages are set to std::nullopt.
+  std::vector<T> min_values_;
+  std::vector<T> max_values_;
+  /// A list of page indices for not-null pages.
+  std::vector<int32_t> non_null_page_indices_;
+};
+
+class OffsetIndexImpl : public OffsetIndex {
+ public:
+  explicit OffsetIndexImpl(const format::OffsetIndex& offset_index) {
+    page_locations_.reserve(offset_index.page_locations.size());
+    for (const auto& page_location : offset_index.page_locations) {
+      page_locations_.emplace_back(PageLocation{page_location.offset,
+                                                page_location.compressed_page_size,
+                                                page_location.first_row_index});
+    }
+  }
+
+  const std::vector<PageLocation>& GetPageLocations() const override {
+    return page_locations_;
+  }
+
+ private:
+  std::vector<PageLocation> page_locations_;
+};
+
+}  // namespace
+
+// ----------------------------------------------------------------------
+// Public factory functions
+
+std::unique_ptr<ColumnIndex> ColumnIndex::Make(const ColumnDescriptor& descr,
+                                               const void* serialized_index,
+                                               uint32_t index_len,
+                                               const ReaderProperties& properties) {
+  format::ColumnIndex column_index;
+  ThriftDeserializer deserializer(properties);
+  deserializer.DeserializeMessage(reinterpret_cast<const uint8_t*>(serialized_index),
+                                  &index_len, &column_index);
+  switch (descr.physical_type()) {
+    case Type::BOOLEAN:
+      return std::make_unique<TypedColumnIndexImpl<BooleanType>>(descr, column_index);
+    case Type::INT32:
+      return std::make_unique<TypedColumnIndexImpl<Int32Type>>(descr, column_index);
+    case Type::INT64:
+      return std::make_unique<TypedColumnIndexImpl<Int64Type>>(descr, column_index);
+    case Type::INT96:
+      return std::make_unique<TypedColumnIndexImpl<Int96Type>>(descr, column_index);
+    case Type::FLOAT:
+      return std::make_unique<TypedColumnIndexImpl<FloatType>>(descr, column_index);
+    case Type::DOUBLE:
+      return std::make_unique<TypedColumnIndexImpl<DoubleType>>(descr, column_index);
+    case Type::BYTE_ARRAY:
+      return std::make_unique<TypedColumnIndexImpl<ByteArrayType>>(descr, column_index);
+    case Type::FIXED_LEN_BYTE_ARRAY:
+      return std::make_unique<TypedColumnIndexImpl<FLBAType>>(descr, column_index);
+    default:
+      break;
+  }
+  DCHECK(false) << "Should not be able to reach this code";
```

Comment:
Can use `arrow::Unreachable()` from `arrow/util/unreachable.h`.

Reply:
Instead can we remove the default from the switch? or at least throw an exception, crashing seems like the wrong choice if a new enum is added.

Reply:
> Instead can we remove the default from the switch? or at least throw an exception, crashing seems like the wrong choice if a new enum is added.

Agreed. In this way, the developer gets noticed there is something to fix at compile time instead of run time.